app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/StockHelper/AddStockStatusToProductsMultistockPlugin.php

```diff
@@ -0,0 +1,94 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Plugin\CatalogInventory\StockHelper;
+
+use Magento\Catalog\Model\Product;
+use Magento\Catalog\Model\ResourceModel\Collection\AbstractCollection;
+use Magento\CatalogInventory\Helper\Stock as Helper;
+use Magento\InventoryApi\Api\Data\StockInterface;
+use Magento\InventoryApi\Api\IsProductInStockInterface;
+use Magento\InventorySalesApi\Api\Data\SalesChannelInterface;
+use Magento\InventorySalesApi\Api\StockResolverInterface;
+use Magento\Store\Api\Data\WebsiteInterface;
+use Magento\Store\Model\StoreManagerInterface;
+
+/**
+ * Plugin for Magento\CatalogInventory\Helper::addStockStatusToProducts.
+ */
+class AddStockStatusToProductsMultistockPlugin
+{
+    /**
+     * @var StockResolverInterface
+     */
+    private $stockResolver;
+
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * @var IsProductInStockInterface
+     */
+    private $isProductInStockInterface;
+
+    /**
+     * @param StockResolverInterface $stockResolver
+     * @param StoreManagerInterface $storeManager
+     * @param IsProductInStockInterface $isProductInStockInterface
+     */
+    public function __construct(
+        StockResolverInterface $stockResolver,
+        StoreManagerInterface $storeManager,
+        IsProductInStockInterface $isProductInStockInterface
+    ) {
+        $this->stockResolver = $stockResolver;
+        $this->storeManager = $storeManager;
+        $this->isProductInStockInterface = $isProductInStockInterface;
+    }
+
+    /**
+     * Around plugin for Magento\CatalogInventory\Helper::addStockStatusToProducts.
+     *
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     * @param Helper $subject
+     * @param callable $proceed
+     * @param AbstractCollection $productCollection
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function aroundAddStockStatusToProducts(
+        Helper $subject,
+        callable $proceed,
+        AbstractCollection $productCollection
+    ) {
+        // We need to prevent the collection modification by plugins next in the chain.
+        $temporaryCollection = clone $productCollection;
```

Comment:
We can not do it
Because it means that functionality in next plugins will be broken (will be applied on clone, but not on real collection)